Cherry pick to correct unintentional merge/unmerge of previous pull r…

Calerid opened the pull request:
Copy and pasted from original pull request: 

I made several small fixed -
Changed the way my round counting system worked.
I. Added new nobodyWins variable for draws. This results in a message stating that no one won the match after 3 draws.
II. Fixed the game never starting a new round.
III. globalReset() can now be called to reset all variables to 0 at the end of a match.
IV. The match would end at 4 with no clear outcome if there wasn't enough draw/player/comp wins. Changed equals or less than. Rounds should always end with a clear outcome.

Removed repeat messages were possible and made them into reusable messages or functions. I could not figure out how to use string/template literals as variables (Any idea?).

Calerid committed:
Cherry pick to correct unintentional merge/unmerge of previous pull request.

PejmanPoh commented on .vscode/launch.json:
```diff
@@ -0,0 +1,15 @@
+            "webRoot": "${workspaceFolder}"
+        }
+    ]
+}
```
You shouldn't commit any files related to .vscode since it is editor specific and other developers on this codebase may be using different editors. To stop files like these from showing up as "unstaged changes", you should use a .gitignore file (https://www.pluralsight.com/guides/how-to-use-gitignore-file).

PejmanPoh commented on My_Odin_Workspace.code-workspace:
```diff
@@ -8,6 +8,6 @@
-		"liveServer.settings.multiRootWorkspaceName": "the_odin_project_projects"
+		"liveServer.settings.multiRootWorkspaceName": "My_Odin_Projects"
 	}
 }
```
Also a VS Code related file

PejmanPoh commented on rock-paper-scissors/index.html:
```diff
@@ -1,11 +1,24 @@
+        </p>
+        <li>Right click; select inspect</li>
+        <li>Select Console</li>
+        <li>enjoy!</li>
```
Haha this is an adorable way of making an interactive site. Get the user to open the console :P Wonder why the tutorial doesn't just have you making a terminal application.

PejmanPoh commented on rock-paper-scissors/scripts/rock-paper-scissors.js:
```diff
@@ -1,13 +1,33 @@
+//Calls the computer play function
+
+computerPlay();
 playGame(playerSelection, computerPlay());
```
In general, functions should be at the top of a file, think of it as a kind of "glossary". The "executable" main code should be right at the bottom so it can use any of the functions above it. This advice won't really apply in real applications (because you won't have code that just executes like this in a real application, everything should be triggered by the user), but I thought it's worth mentioning. Might give you an idea of what should go where.

PejmanPoh commented on rock-paper-scissors/scripts/rock-paper-scissors.js:
```diff
@@ -1,13 +1,33 @@
 let playerWins = 0;
 let computerWins = 0;
+let nobodyWins = 0;
 let roundCount = 0;
```
I think the better alternative of what I am about to say is a bit beyond scope atm but I want to plant this seed early. These are global variables (because they can be accessed by everything in this file) and they are evil. The reason has to do with the term "IMMUTABILITY" but in general global variables are breeding grounds for bugs. Very hard to track down bugs. Don't worry about it for now though.

PejmanPoh commented on rock-paper-scissors/scripts/rock-paper-scissors.js:
```diff
@@ -1,13 +1,33 @@
+let computerSelection = "";
+
+function computerPlay() {
+  let rPSS = ["rock", "paper", "scissors"];
```
A better name for this variable could be "options" or something. The reason being if you wanted to add a new type like "lizard" or "spock" (https://www.wikihow.com/Play-Rock-Paper-Scissors-Lizard-Spock), you don't need to rename the variable. The general rule behind this is "name the collection, not the items".

It could also work better outside the function as a constant (different from global variables, as constants in theory don't change during runtime).

PejmanPoh commented on rock-paper-scissors/scripts/rock-paper-scissors.js:
```diff
@@ -1,13 +1,33 @@
+
 function startNewRound() {
-  playerSelection = prompt("Please enter Rock, Paper, or scissors");
+  playerSelection = prompt(
+    "Please enter Rock, Paper, or scissors"
+  ).toLowerCase();
```
So there is a concept called "DRY" (https://en.wikipedia.org/wiki/Don%27t_repeat_yourself). Don't need to read the wikipedia fully but the idea behind it is, if you've got the same code in two places, you can probably simplify it.

In this case you've got l5 and l28 doing the same thing. What you can do instead is create a function called:
```
function getPlayerSelection() {
    return prompt(
        "Please enter Rock, Paper, or scissors"
    ).toLowerCase();
}
```

Then whenever you need to get a player's selection (such as here on l28) you type:
`let playerSelection = getPlayerSelection()`.

The benefits of doing this:
1. Not repeating yourself means that if you changed this code to include "lizard" and "spock", you don't have to replace multiple lines (or worse, forget to do so and have a bug!)
2. You can easily get rid of the (evil) global variable on l5. Instead, you can just have a `let playerSelection = getPlayerSelection()` on l35 to compensate.
3. `let playerSelection = getPlayerSelection()` is more readable than the prompt and then `toLowerCase()`.

PejmanPoh replied:
A further optimisation, if you have the player selection occur right at the beginning of `playGame()`, you don't need it here on l28. 

ANOTHER benefit of this is you don't have to pass in `playerSelection`, reducing the overall number of variables being passed around (less bugs!).

PejmanPoh commented on rock-paper-scissors/scripts/rock-paper-scissors.js:
```diff
@@ -21,89 +41,110 @@ function playGame(playerSelection, computerSelection) {
-
-function roundCounter(roundCount) {}
 
 function rock(computerSelection) {
```
There is a LOT of repetition going on between `rock()`, `paper()`, and `scissors()`. I'll put some general optimisations in th thread below.

PejmanPoh replied:
Instead of
```
roundCount += 1;
playerWins += 1;
```
being repeated in all three functions, you can write a function:
```
function calculatePlayerWins() {
    roundCount += 1;
    playerWins += 1;
}
```

PejmanPoh replied:
Adding the console.logs into the above functions, you'll need to add the `playerSelection` and `computerSelection` as variables.

So it becomes:
```
function calculatePlayerWins(playerSelection, computerSelection) {
    roundCount += 1;
    playerWins += 1;
    console.log(`You win! ${playerSelection} beats ${computerSelection}`);
}
```

PejmanPoh replied:
Once you've got all 3 functions:
```
function calculateNobodyWins() {
    roundCount += 1;
    nobodyWins += 1;
    console.log(`This round was a draw! Try again!`);
}

function calculatePlayerWins(playerSelection, computerSelection) {
    roundCount += 1;
    playerWins += 1;
    console.log(`You win! ${playerSelection} beats ${computerSelection}`);
}

function calculateComputerWins(playerSelection, computerSelection) {
    roundCount += 1;
    computerWins += 1;
    console.log(``You Lose! ${computerSelection} beats ${playerSelection}!`);
}
```

There is one more (obvious) point of repetition. Can you spot it? It's to do with `roundCount`.

PejmanPoh commented on rock-paper-scissors/scripts/rock-paper-scissors.js:
```diff
@@ -21,89 +41,110 @@ function playGame(playerSelection, computerSelection) {
   return roundCounter(roundCount);
 }
 
-function roundCOunter(roundCount) {
```
Using the lessons from above, try to optimise this function to remove repetition (use lots of functions!).

PejmanPoh commented on rock-paper-scissors/scripts/rock-paper-scissors.js:
```diff
@@ -21,89 +41,110 @@ function playGame(playerSelection, computerSelection) {
+  computerWins = 0;
+  nobodyWins = 0;
+  roundCount = 0;
+  return console.log("score reset");
```
This return is unnecessary. You can just `console.log`.

PejmanPoh reviewed (changes requested):
General comments:
1. I will be commenting based on the idea that there would be other developers in the codebase and you are writing code for others to read as well. This is the basic idea behind coding. Making it readable for other developers.
2. You can read my comments in chronological order, I build on some comments off previous ones.
3. I foresee potential bugs because there is a lot of repetition and with it comes mistaken logic (e.g. setting the wrong winner based on a complicated tree of logic).

It all looks good though and it shows good foundational knowledge! There's definitely room for improvement but you can only build a foundation one brick at a time.

Calerid committed:
Deleted file specific to local machine

File was specific to vscode on local machines. Deleted the file as it was unnecessary to repository.

Calerid committed:
Deleted launch.json specific to local machine


Deleted the file which was specific to a local machine. Not pertinent to the repository.

Calerid committed:
Updated README to include new project


Added the Rock Paper Scissors project to the readme file

Calerid committed:
Removed rPSS from computerPlay

Removed rPSS as a variable contained inside of computerPlay, made it a const variable in the variable section.

Calerid committed:
Created function playerPlay

Created a new function which contains playerSelection. Attempting to reduce redundant code

Calerid committed:
Moved startNewRound() up, added comments on code section

Moved the startNewRound function up. After eliminating unecessary calls to other functions, added startNewRound() call to initiate the script as a whole. Additionally added some comments to assist with making the program more readable.

Calerid committed:
Created three new functions for adding to counting variables

Created the functions addPlayerWins, addComputerWins, and addNobodyWins to help eliminate redundancy within rock(), paper(), & scissors()

Calerid committed:
Added 'add' player, comp, and nobodyWins to rock, paper, & scissors

Removed the repeated variable '+=' to now call individual functions for each of the outcomes.
This was done to reduce redundant code

Calerid committed:
Added comments to improve readability of code

Added comments to different sections indicating what each group of functions performs in the program. This was done to improve readability

Calerid committed:
Moved variables computerSelection & playerSelection

Moved the variables into startNewRound(). This was done to make their scope local and prevent future errors caused by global variables

Calerid committed:
Added arguments to add functions, moved original variabes to global

Added the variables computerSelection and playerSelection to the 'add' functions. Also returned the two variables to global from startNewRound(),

Calerid committed:
Add function newGamePrompt()

Add newGamePrompt for reduced redundance in roundCounter(). yes/no input.

Calerid committed:
Add newGamePrompt() to roundCounter

Change roundCounter to call newGamePrompt(). Remove reference to yN, now playAgain.

Calerid committed:
Add function newGameNo()

Function add noNewGame(). console.log thank you message, score. Run globalReset()

Calerid committed:
Fix newGamePrompt / playAgain in rouncCounter

Fixed newGamePrompt passing in variable playAgain into roundCounter(). Now passes corect variable

Calerid committed:
Add newGameNo() in roundCounter()

Added the function newGameNo() into roundCounter() when the option playAgain === No. This will print the match score and reset the global variables

Calerid committed:
Add newGameYes() to roundCounter()

Added the function newGameYes() to the function roundCounter(). This will reduce redundant calls for globalReset() and startNewRound().

Calerid committed:
Fix indent issue with roundCounter()

Fixed an issue with the indentation in the function RoundCounter(). This was causing all rounds to result in a draw, ignoring the count of player and computer wins.

PejmanPoh commented on rock-paper-scissors/scripts/rock-paper-scissors.js:
```diff
@@ -0,0 +1,158 @@
+}
+
+// "add" functions tally the score. called by functions rock(), paper(), scissors().
+function addPlayerWins(playerSelection, computerSelection) {
+  console.log(`You win! ${playerSelection} beats ${computerSelection}`);
+  roundCount += 1;
+  playerWins += 1;
+}
+function addComputerWins(playerSelection, computerSelection) {
+  console.log(`You Lose! ${computerSelection} beats ${playerSelection}!`);
+  roundCount += 1;
+  computerWins += 1;
+}
+function addNobodyWins() {
+  console.log(`This round was a draw! Try again!`);
+  roundCount += 1;
+  nobodyWins += 1;
```
So the way I alluded to simplifying these three functions is by noticing `roundCount += 1` is a piece of code that is repeated, regardless of who won. That means we can likely extract it (like a mathematical equation) and put it outside. i.e. `xa + xb + xc == x(a + b + c)` 

If you move `roundCount += 1` to l68, l79, and l90. You can remove some of the duplication. The fact that it still needs to be in three places indicates we can simplify further, but I think it's good enough for now.

PejmanPoh commented on rock-paper-scissors/scripts/rock-paper-scissors.js:
```diff
@@ -0,0 +1,158 @@
+// the player and computer functions to make a selection.
+
+startNewRound();
+function startNewRound() {
+  playerPlay();
+  computerPlay();
+  playGame(playerSelection, computerSelection);
+}
```
So in an effort to reduce the number of global variables, what you were missing the last time we spoke was l15 and l16. They need to be:

```
  const playerSelection = playerPlay();
  const computerSelection = computerPlay();
```

With l2 and l3 gone. And the `playerPlay()` functions need to be:
```
function playerPlay() {
    return prompt("Please enter Rock, Paper, or scissors").toLowerCase();
}
```

The flow here is we store the result of the `playerPlay()` function into the new local variable `playerSelection` and pass that along.

Calerid replied:
Some confusion here. With this it seems that because the variables are constant, I cannot reassign them on new rounds. So whatever my input is on the first round, it remains during following rounds.

PejmanPoh replied:
Declaring `const`s inside the function make them local scoped. So any new call of `startNewRound()` will reinitialise all of the variables, and require prompts. The code should look like this:

```
function playerPlay() {
    return prompt("Please enter Rock, Paper, or scissors").toLowerCase();
}

function startNewRound() {
  const playerSelection = playerPlay();
  const computerSelection = computerPlay();
  playGame(playerSelection, computerSelection);
}
```

Does that not work?

PejmanPoh commented on rock-paper-scissors/scripts/rock-paper-scissors.js:
```diff
@@ -0,0 +1,158 @@
+  if (roundCount < 5) {
+    if (playerWins === 3) {
+      console.log("You are the winner!");
+      playAgain = newGamePrompt();
+      if (playAgain === "yes") {
+        newGameYes()
+      } else if (playAgain === "no") {
+        newGameNo()
+      }
```
As with a comment above, it's useful to treat `if` statements like mathematical equations. We can see that this block of code is repeated 3 times. Already we can extract that to a function to make it easier to read e.g.
```
function playAgain() {
    const playAgain = newGamePrompt();
      if (playAgain === "yes") {
        newGameYes()
      } else if (playAgain === "no") {
        newGameNo()
      }
}
```

But then we are left with
```
 if (playerWins === 3) {
      console.log("You are the winner!");
      playAgain();
} else if  (computerWins === 3) {
       console.log("You lose!");
       playAgain();
} else if (nobodyWins === 3) {
        console.log("Nobody won!");
       playAgain();
} else {
        startNewRound();
}
```

And I think we can try to figure out a way to extract `playAgain()` from this just like a mathematical equation. I'll leave that up as an exercise for you to figure out though! Hint: You'll likely need a new wrapping `if` statement.

As an aside, I feel like this code can't work? `playAgain` isn't declared anywhere in the scope of this function?

PejmanPoh commented on rock-paper-scissors/scripts/rock-paper-scissors.js:
```diff
@@ -0,0 +1,158 @@
+const gameChoices = ["rock", "paper", "scissors"];
```
FYI: Constants tend to be in singular capitalised snake case e.g. `GAME_CHOICE`

Singular because then when you select one of them (e.g. `GAME_CHOICE[1]`), you read it as "Game Choice 1"

PejmanPoh reviewed (commented):
I have a few more things I think we can work on after this, I think `roundCounter()` has a lot of interesting things about it. But we can work on that after you address the current issues. They aren't that easy to figure out so you'll need a bit of thought.

Calerid committed:
Move roundCount to startNewRound()

Moved 'roundCount += 1' from addComputerWins into startNewRound() to reduce redundant code.

Calerid committed:
Make const gameChoices snake case

Changed the constant into gameChoices into snake case GAME_CHOICE. All instances should be changed now.

Calerid committed:
Add startNewRound() button

Created a button with the trigger onclick in index.html attached to a button 'play'. This was done to make the game easier to play, as well as test. Button highlighting was removed. Instructions were also added/altered under the index.html file to make it reflect the changes, this will hopefully make the game more user friendly when complete.

Calerid committed:
Move console.log() stats

Moved the console.log of game stats to roundCounter. This will not print the stats of the game every single round instead of on completion of a match. This was done to see why some rounds would end with a draw when player or computer reached 3 wins. Will leave so player can see what score is at each round

Calerid committed:
Add playerwins computer wins to several functions/calls

Added playerWins and computerWins to several functions and their calls/returns. Somehow these were being passed to the functions anyways; however, this was to prevent errors that may occur from them not being passed.

Calerid committed:
Add error handling to playgame()

Added an else if statement and else statement to playGame. Chrome has a violation warning which occurs when the function takes too long to play, this stops the game from counting correctly. Else if takes the input of start, this will be added to index.html later.

Else statement added to catch input in prompt which are not part of the game. If a player enters an invalid string, the game will remove round from count and run startNewRound(). This prevents the game from crashing on an invalid string.

Calerid committed:
Add error handling for rouncount > 5

Added an else if statement roundCounter which detects if roundCount is greater than five. If this is the case, a new game is triggered. This was added because the game should always end at five rounds.

Calerid committed:
Extract playagain to new function

Created a new function to reduce redundant code in roundCounter().
Now calls on newGamePrompt(). This has created an issue with an event handler timeout. This commit may be reversed if a fix is not found.

Calerid commented:
That's probably the issue. I was attempting to declare them globally. I'll
try when I can and see. I have another issue to fix with the script timing
out first, then I'll get to that.

On Fri, Jan 15, 2021, 4:23 AM Pejman Poh <notifications@github.com> wrote:

> *@PejmanPoh* commented on this pull request.
> ------------------------------
>
> In rock-paper-scissors/scripts/rock-paper-scissors.js
> <https://github.com/Calerid/My_Odin_Projects/pull/3#discussion_r558203355>
> :
>
> > +function startNewRound() {
> +  playerPlay();
> +  computerPlay();
> +  playGame(playerSelection, computerSelection);
> +}
>
> Declaring consts inside the function make them local scoped. So any new
> call of startNewRound() will reinitialise all of the variables, and
> require prompts. The code should look like this:
>
> function playerPlay() {
>     return prompt("Please enter Rock, Paper, or scissors").toLowerCase();
> }
>
> function startNewRound() {
>   const playerSelection = playerPlay();
>   const computerSelection = computerPlay();
>   playGame(playerSelection, computerSelection);
> }
>
> Does that not work?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Calerid/My_Odin_Projects/pull/3#discussion_r558203355>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHHBG3K3HBRPIEYD6TEFNNDS2AJTFANCNFSM4V3RLPYQ>
> .
>

PejmanPoh reviewed (approved):
I see a ton of improvement already on the codebase. You have really reduced the repetitions and I think that's a huge takeaway already from this exercise. There's always things to be improved but I think it's honestly good enough. If it compiles and works bug-free, I think we can follow the Pareto Principle and say you've done your 20%. I'm happy if you move along. Of course I'm also fine if you still want to improve this code.

Calerid committed:
Fix roundcounter()

Added an else statement that calls a new round if win criterea are not met. This was ommited during the refactor / extraction of newGamePrompt() and caused the game to run only once when the button was pressed.

Calerid commented:
Excellent, I concur. Made one fix to roundCounter(). Now making the merge!
